dev/Breeze/requirements.txt

```diff
@@ -0,0 +1,4 @@
+click
+pytest
+pytest-xdist
+rich
```

Comment:
We might consider that indeed.

I think there is a certain simplicity that the code to manage development is inside the repo with the code. It solves a number of problems, and especially mkes it easier to use versioning. I would treat that as an option to follow in the future and the way how we "release it" eventually when we get to the point where it is is "ready to use". 

One potential problem with it is that you have to actively "install" the latest version if you want new features and by having everything in the same repo, of the code that is used to manage it, you have it "by default" (which also means that  it is less stable). 

Keeping it under "dev/Breeze" folder does not preclude having a separate package and PyPI release. There could be two modes of runnign Breeze - one from the sources, and one from the package.

Reply:
We should make an ADR at some point about it at a later phase weighting pros/cons and generally deciding about the "distribution" methonds

Reply:
> One potential problem with it is that you have to actively "install" the latest version if you want new features and by having everything in the same repo

I mentioned keeping the `Breeze` command small so this won’t be the case. The command will mostly just contain

```python
def _find_airflow_repo_root():
    ...

if __name__ == "__main__":
    import sys

    root = _find_airflow_repo_root()
    if root is None:
        # emit some sort of error message...
        sys.exit(1)
    sys.path.insert(0, root / "dev")

    import breeze
    breeze.run()
```

and an upgrade is only needed if we need to change dependencies, which should be seldom (and when it happens you need to pip-install an in-tree Breeze’s requirements.txt anyway).

Reply:
Ah I see. I misunderstood that. Hmm. Might be a good idea for kind of "bootstraping" of breeze. I like the idea.

Reply:
Still would not help with running tests via CI where conftest.py is there mostly to overcome default import behaviour of `pytest` when it comes to importing. The funny thing is that `python -m pytest` works fine, but when you try to run tests with `pytest`, the imports are not working the way you'd think.

Reply:
Because `python -m` adds the current working directory to `sys.path`, while an entry point (which is a script) does not. This is one of the lesser known Python quirk; I don’t know what the actual name is but personally call it “interpreter mode” (which adds cwd) and “script mode” (which does not).

Pytest actually has specific documentation on this: https://docs.pytest.org/en/6.2.x/goodpractices.html#choosing-a-test-layout-import-rules

This is why having a pip-installable package helps; we can lay out `dev/breeze` as a Python package:

```
airflow/dev/breeze
 |-- pyproject.toml
 |-- setup.py  # assuming we’ll use setuptools...
 |-- src/
 |    `-- breeze/  # importable module
 `-- tests/
      |-- conftest.py
      `-- test_breeze.py
```

And `pip install -e ./dev/breeze` + `pytest dev/breeze/tests` would work automatically.

Reply:
So - what you suggest is to have "pip install -e ./dev/breeze` in a virtualenv? I like it.

What I really want (as end-goal) is something that we have with the old breeze where in order to use it, you do not worry about having a separate venv specifically prepared manually by the user (it should be created automatically in ~/.build/ folder and activated when breeze commands are run. 

All the installation should happen under-the-hood and you should be able to run both breeze and tests without having to switch to that venv (kind of what pipx does). And autocomplete should also work with the command.

Do you think that would work in this case (I think so) with the "package" structure you propose ?

Reply:
OK. Updated it to follow your suggestions @uranusjr - I like it much more too. Can you please take a look ? Setuptools and configuration of it is kinda bad documented so some of the stuff there is a bit guessing.